crates/symbolicator/src/services/cacher.rs

```diff
@@ -277,51 +295,93 @@ impl<T: CacheItemRequest> Cacher<T> {
+            name: self.config.name(),
+            version: T::VERSIONS.current,
+            local_key: key.clone(),
+        };
+        let dup_file = temp_file.as_file().try_clone()?;
+        let mut temp_fd = tokio::fs::File::from_std(dup_file);
+
+        // TODO: consider cache expiry!
+        let shared_cache_hit = self
+            .shared_cache_service
+            .fetch(&shared_cache_key, &mut temp_fd)
+            .await;
+
+        let status = if !shared_cache_hit {
+            let status = request.compute(temp_file.path()).await?;
+            status.write_marker(&mut temp_fd).await?;
+            Some(status)
+        } else {
+            None
+        };
 
-        let status = request.compute(temp_file.path()).await?;
+        // Now we have written the data to the tempfile we can mmap it, persisting it later
+        // is fine as it does not move filesystem boundaries there.
+        let byte_view = ByteView::map_file_ref(temp_file.as_file())?;
 
-        if let Some(ref cache_path) = cache_path {
-            sentry::configure_scope(|scope| {
-                scope.set_extra(
-                    &format!("cache.{}.cache_path", name),
-                    cache_path.to_string_lossy().into(),
-                );
-            });
+        let status = status.unwrap_or_else(|| CacheStatus::from_content(&byte_view));
 
-            log::trace!("Creating {} at path {:?}", name, cache_path);
-        }
+        let path = match self.config.cache_dir() {
+            Some(ref cache_dir) => {
+                // Cache is enabled, write it!
+                let cache_path = key.cache_path(cache_dir, T::VERSIONS.current);
 
-        let byteview = ByteView::open(temp_file.path())?;
-
-        metric!(
-            counter(&format!("caches.{}.file.write", name)) += 1,
-            "status" => status.as_ref(),
-            "is_refresh" => &is_refresh.to_string(),
-        );
-        metric!(
-            time_raw(&format!("caches.{}.file.size", name)) = byteview.len() as u64,
-            "hit" => "false",
-            "is_refresh" => &is_refresh.to_string(),
-        );
-
-        let path = match cache_path {
-            Some(ref cache_path) => {
-                status.persist_item(cache_path, temp_file)?;
-                CachePath::Cached(cache_path.to_path_buf())
+                log::trace!(
+                    "Creating {} at path {:?}",
+                    self.config.name(),
+                    cache_path.display()
+                );
+                let parent = cache_path.parent().ok_or_else(|| {
+                    std::io::Error::new(
+                        std::io::ErrorKind::Other,
+                        "no parent directory to persist item",
+                    )
+                })?;
+                fs::create_dir_all(parent).await?;
+                let mut file = temp_file.persist(&cache_path).map_err(|x| x.error)?;
+
+                let file_size = file.seek(SeekFrom::End(0))?;
```

Comment:
we have the `byte_view` already from above which we can use like before. why are we seeking here?

Reply:
oh, fair enough.  at some point there wasn't a byteview yet here.